ipykernel/kernelbase.py

```diff
@@ -1007,13 +1021,35 @@ def do_is_complete(self, code):
         return {"status": "unknown"}
 
     async def debug_request(self, socket, ident, parent):
-        """Handle a debug request."""
+        """Handle a debug request.
+
+        Seem Deprecated
+        """
+        # If this is incorrect, remove `debug_request` from the deprecated message types
+        # at the beginning of this class
+        warnings.warn(
+            "debug_request is deprecated in kernel_base since ipykernel 6.10"
+            " (2022). It is only part of IPython parallel. Did you mean do_debug_request ?",
+            DeprecationWarning,
+            stacklevel=2,
+        )
         if not self.session:
             return
         content = parent["content"]
-        reply_content = self.do_debug_request(content)
-        if inspect.isawaitable(reply_content):
-            reply_content = await reply_content
+        if not inspect.iscoroutinefunction(self.do_debug_request):
+            # repeat the warning at run
+            reply_content = self.do_debug_request(content)
+            warnings.warn(
+                "`do_debug_request` will be required to be a coroutine "
+                "functions in the future. coroutine functions have ben supported "
+                "since ipykernel 6.0 (2021)",
+                DeprecationWarning,
+                stacklevel=1,
+            )
+            if inspect.isawaitable(reply_content):
```

Comment:
Why not throwing the warning if `reply_content` is not awaitable?

Reply:
Because testing for awaitable is/was the incorrect things to do.

I believe at some point sync functions were returning Futures, which are awaitable. 

We do not want the return a `do_debug_request` to be an awaitble; we want `do_debug_request` to be a coroutine function.

Reply:
There are good reasons to return an awaitable rather than using a coroutine (it can have dramatic performance benefits), and cutting that off due to deficiencies in tooling does not seem like a positive change to me.

mypy [appears to understand](https://mypy-play.net/?mypy=1.11.0&python=3.10&gist=5ab30e161d5e83a16538d790c9c2deb9) `inspect.isawaitable` since 1.11, so this check is correctly handled for type narrowing, but we can also use `isinstance(thing, typing.Awaitable)`, which is equivalent and I think has worked in mypy for longer.

Reply:
Fine; we can leave it is awaitable; but should we still support both sync and async ?

Reply:
`do_xyz` methods were sync-only for a long time, so most implementations are sync because they have no need to be async. It costs us nothing to keep it that I can see, and it breaks users for no benefit, so I don't understand the trade off. In general, I think you need a really good reason to break APIs, and version numbers don't really have any effect on that.

Reply:
> It costs us nothing to keep it that I can see, and it breaks users for no benefit, so 
I don't understand the trade off

It never "cost nothing" to keep optional things like this – it's not the first time you've used this justification, I used to agree, but I've grown.

It cost not much, but on a really long tail – and it accumulates. Keeping it means that every other consumer present or future of `do_debug_request` must keep in mind that it might be sync or async, and test for it – meaning that `await do_debug_request(...)` is technically wrong at it may be sync and consumer then _must_ check for `isawaitable(result)` and await instead.

This also make forgetting an `async` in subclass a subtly incorrect but working implementations:

```
class Sup:

    async def do_debug(self):
        return 'res'

class Derived(Sup):

    def do_debug(self):
        print('Log')
        return super().do_debug()
```

Well you see the kind of issues.

The non-awaitable/non async path is BTW not tested, and the test in this repo assume the method is always a coroutine:

```
tests/test_ipkernel_direct.py
208:async def test_do_debug_request(ipkernel: IPythonKernel) -> None:
211:    await ipkernel.do_debug_request(msg)


tests/conftest.py
106:    async def do_debug_request(self, msg):
```

Personally I got bitten too many time by "it costs nothing" (think shim modules of IPython, and ipython_genutils and alike from a decade ago), that are still regularly used in client projects giving me headache because we should have ripped them out earlier instead of other folks (or us in the future) paying costs for a long time.

Same things with the it can be X or Y  "for convenience" (Api with strings or bytes, $element or iterable of $element where a consumer ends up passing $element which is iterable and tnigs break,... etc) where you always end up chasing a bug for hours and cursing the API.

Hey up until last week I was chasing a bug in reformatting code in darker,  because In IPython we have a test case in IPython for cyrillic, due to some lingering `py3compat.cast_unicode` in IPython ...

And again, I'm not asking to break now; I'm suggesting we tell consumer "hey this is a small but long term cost we will likely not want to incur forever"

Reply:
I don't agree with your assessment of the trade-offs in this case and think it shifts more cost to users than is appropriate, but since you feel strongly, if you want to keep the deprecation message for `if not isawaitable(result)` I can accept that.

I do _super_ appreciate your diligence in adding "since when" info to deprecation messages. That has been a huge help for me.

Once all the deprecations about message handler functions are removed (the actually deprecated messages are now gone, debug_request is part of the [standard kernel protocol](https://jupyter-client.readthedocs.io/en/latest/messaging.html#debug-request), not deprecated or related to ipyparallel), this should be all set.

Reply:
> I don't agree with your assessment of the trade-offs in this case and think it shifts more cost to users than is appropriate, but since you feel strongly, if you want to keep the deprecation message for `if not isawaitable(result)` I can accept that.

Would a PendingDeprecationWarning, or a FutureWarning with a softer message ease your concerns ? saying that we are _considering_ making it mandatory, and that we would _prefer_ implementer to always return an awaitable ?



> I do _super_ appreciate your diligence in adding "since when" info to deprecation messages. That has been a huge help for me.

Thanks, it feel so frustrating when message don't give you this information and just give you the stick with "this will be removed in..."; Maybe one day I'll give a talk on why/how you should do that and other things in deprecation warnings and how to make it easier to push users for new API; or maybe just an blog post ? idk.



> Once all the deprecations about message handler functions are removed (the actually deprecated messages are now gone, debug_request is part of the [standard kernel protocol](https://jupyter-client.readthedocs.io/en/latest/messaging.html#debug-request), not deprecated or related to ipyparallel), this should be all set.

I need to redo this PR anyway, some of the deprecation were due to improper comments and documentation; I'll try to get things in a more correct state; and maybe split that into multiple PR

Reply:
I think DeprecationWarning or PendingDeprecationWarning is fine, either way.